File: runtimes/java/versions/latest/src/main/java/io/openruntimes/java/RuntimeResponse.java
package io.openruntimes.java;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import java.util.HashMap;
import java.util.Map;

public class RuntimeResponse {
  private static final Gson gson = new GsonBuilder().serializeNulls().create();

  public RuntimeOutput binary(byte[] bytes, int statusCode, Map<String, String> headers) {
    return new RuntimeOutput(bytes, statusCode, headers);
  }

  public RuntimeOutput binary(byte[] body, int statusCode) {
    return this.binary(body, statusCode, new HashMap<>());
  }

  public RuntimeOutput binary(byte[] body) {
    return this.binary(body, 200, new HashMap<>());
  }

  public RuntimeOutput send(String body, int statusCode, Map<String, String> headers) {
    return this.text(body, statusCode, headers);
  }

  public RuntimeOutput send(String body, int statusCode) {
    return this.send(body, statusCode, new HashMap<>());
  }

  public RuntimeOutput send(String body) {
    return this.send(body, 200, new HashMap<>());
  }

  public RuntimeOutput text(String body, int statusCode, Map<String, String> headers) {
    return this.binary(body.getBytes(), statusCode, headers);
  }

  public RuntimeOutput text(String body, int statusCode) {
    return this.text(body, statusCode, new HashMap<>());
  }

  public RuntimeOutput text(String body) {
    return this.text(body, 200, new HashMap<>());
  }

  public RuntimeOutput json(Map<String, Object> json, int statusCode, Map<String, String> headers) {
    headers.put("content-type", "application/json");
    return this.text(gson.toJson(json), statusCode, headers);
  }

  public RuntimeOutput json(Map<String, Object> json, int statusCode) {
    return this.json(json, statusCode, new HashMap<>());
  }

  public RuntimeOutput json(Map<String, Object> json) {
    return this.json(json, 200, new HashMap<>());
  }

  public RuntimeOutput empty() {
    return this.text("", 204, new HashMap<>());
  }

  public RuntimeOutput redirect(String url, int statusCode, Map<String, String> headers) {
    headers.put("location", url);
    return this.text("", statusCode, headers);
  }

  public RuntimeOutput redirect(String url, int statusCode) {
    return this.redirect(url, statusCode, new HashMap<>());
  }

  public RuntimeOutput redirect(String url) {
    return this.redirect(url, 301, new HashMap<>());
  }
}

File: runtimes/kotlin/versions/latest/src/main/kotlin/io/openruntimes/kotlin/RuntimeResponse.kt
package io.openruntimes.kotlin

public class RuntimeResponse {
    fun binary(
        bytes: ByteArray,
        statusCode: Int = 200,
        headers: Map<String, String> = mapOf(),
    ): RuntimeOutput = RuntimeOutput(bytes, statusCode, headers)

    fun text(
        body: String,
        statusCode: Int = 200,
        headers: Map<String, String> = mapOf(),
    ): RuntimeOutput = binary(body.toByteArray(), statusCode, headers)

    fun send(
        body: String,
        statusCode: Int = 200,
        headers: Map<String, String> = mapOf(),
    ): RuntimeOutput = text(body, statusCode, headers)

    fun json(
        json: MutableMap<String, Any>,
        statusCode: Int = 200,
        headers: MutableMap<String, String> = mutableMapOf(),
    ): RuntimeOutput {
        headers["content-type"] = "application/json"
        return this.text(gsonInternal.toJson(json), statusCode, headers)
    }

    fun empty(): RuntimeOutput = this.text("", 204, mutableMapOf<String, String>())

    fun redirect(
        url: String,
        statusCode: Int = 301,
        headers: MutableMap<String, String> = mutableMapOf(),
    ): RuntimeOutput {
        headers["location"] = url
        return this.text("", statusCode, headers)
    }
}
